### PR TITLE
Major MMF2/PAM updates

### LINKED ISSUE #6363: build fails with `SMS_Ln3.ne4pg2_ne4pg2.F2010-MMF2`
Likely due to a change in cmake made in scream repo that was merged into E3SM.

`SMS_Ln3.ne4pg2_ne4pg2.F2010-MMF2.pm-cpu_gnu` (as well as intel) fails to build:

```
-- Configuring done
CMake Error at eam/src/physics/crm/pam/external/physics/scream_cxx_p3_shoc/scream_share/CMakeLists.txt:35 (add_library):
  Cannot find source file:

    /global/cfs/cdirs/e3sm/ndk/repos/me26-apr22/components/eamxx/src/share/grid/remap/horizontal_remap_utility.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc


CMake Error at eam/src/physics/crm/pam/external/physics/scream_cxx_p3_shoc/scream_share/CMakeLists.txt:35 (add_library):
  No SOURCES given to target: scream_share


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

### PR DESCRIPTION
**Pull request by whannah1**
This brings in several updates and fixes for the PAM CRM used in the MMF2 compsets to improve stability and throughput. Some of these required updates to the PAM submodule, but the most important change was the addition of adaptive sub-cycling in the PAM driver. This change allowed the CRM time step to be increased to 10 seconds, which dramatically improved the throughput while also making the model much more resilient to numerically unstable conditions due to unrealistically strong velocities (still not sure why these occur in the first place).

List of notable changes for PAM and/or the driver: 
- adaptive subcycling in PAM driver
- enable variance transport by default
- disable mean-state acceleration for dry density
- use crm_nx=45 with exact interpolation for PAM dycor (requires odd CRM columns)
- increase PAM hyperdiffusion timescale to 60 seconds to reduce over-smoothing

Fixes #6363

[non-BFB] only for MMF2

### CODE REVIEW

**Comment by whannah1** [comment]
The plot below shows various 1-year averages from a recent simulation on OLCF Summit. This run had an average throughput of 1.57 sypd using 128 nodes. 

Note that the liquid and ice water paths fields show a substantial imbalance of too little liquid and too much ice. This needs to be addressed with tuning parameters in the future. The main parameters we want to target are not available to be modified from the MMF/PAM code until the EAMXX fork is pulled into the E3SM repo. So tuning changes are expected to be the focus of a future PR. 

image

**Comment by whannah1** [comment]
The latest commit should address issue #6363

File: components/eam/src/physics/crm/pam/pam_driver.cpp [hunk 1]
```diff
@@ -14,26 +13,90 @@
...
+  //------------------------------------------------------------------------------------------------
+  // initialize max U and W arrays
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    wvel_max(k,n) = 0.0;
+    uvel_max(k,n) = 0.0;
+  });
+  // calculate max U and W
+  parallel_for( SimpleBounds<4>(crm_nz,crm_ny,crm_nx,nens) , YAKL_LAMBDA (int k, int j, int i, int n) {
+    yakl::atomicMax(uvel_max(k,n), sqrt(uvel(k,j,i,n)*uvel(k,j,i,n)) );
+    yakl::atomicMax(wvel_max(k,n), fabs(wvel(k,j,i,n)) );
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_driver.cpp:65]: I prefer `std::abs` to `fabs`

File: components/eam/src/physics/crm/pam/pam_driver.cpp [hunk 2]
```diff
@@ -14,26 +13,90 @@
...
+  real2d cfl_max("cfl_max",  crm_nz,nens);
+  //------------------------------------------------------------------------------------------------
+  // initialize max U and W arrays
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    wvel_max(k,n) = 0.0;
+    uvel_max(k,n) = 0.0;
+  });
+  // calculate max U and W
+  parallel_for( SimpleBounds<4>(crm_nz,crm_ny,crm_nx,nens) , YAKL_LAMBDA (int k, int j, int i, int n) {
+    yakl::atomicMax(uvel_max(k,n), sqrt(uvel(k,j,i,n)*uvel(k,j,i,n)) );
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_driver.cpp:64]: Why not `std::abs` here?

File: components/eam/src/physics/crm/pam/pam_driver.cpp [hunk 3]
```diff
@@ -14,26 +13,90 @@
...
+  // find max CFL between horizontal and vertical CFL values
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    int k_gcm = gcm_nlev-1-k;
+    real crm_dz = input_zint(k_gcm,n) - input_zint(k_gcm+1,n);
+    real cfl_u = uvel_max(k,n)*crm_dt_fixed/crm_dx;
+    real cfl_w = wvel_max(k,n)*crm_dt_fixed/crm_dz;
+    cfl_max(k,n) = max(cfl_u,cfl_w);
+  });
+  // calculate final CFL across ensemble
+  real cfl_loc = pmax(cfl_max.data());
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_driver.cpp:76]: You can now say the following here:
```C++
real cfl_loc = yakl::intrinsics::maxval(cfl_max);
```
Much clearer and easier :)

File: components/eam/src/physics/crm/pam/pam_driver.cpp [hunk 4]
```diff
@@ -14,26 +13,90 @@
...
+  num_subcycle = max(num_subcycle,max(1,static_cast<int>(ceil(cfl/0.7))));
+  real crm_dt_subcycle = crm_dt_fixed / num_subcycle;
+  coupler.set_option<real>("crm_dt",crm_dt_subcycle);
+  // check for excessive subcylcing - don't exit, just print
+  if(num_subcycle > max_num_subcycle) {
+    real umax = pmax(uvel_max.data());
+    real wmax = pmax(wvel_max.data());
+    printf("PAM_DRIVER - WARNING: excessive subcycling!"
+          " - num_subcycle: %3.3d  dt: %8.4f  cfl: %8.4f  umax: %8.2f  wmax: %8.2f \n",
+          num_subcycle,crm_dt_subcycle,cfl,umax,wmax);
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_driver.cpp:88]: You can `std::cout << std::endl` at the end to force the buffer to flush to file / terminal. I like to do that at the end of all warning or error messages to make it much more likely it gets output before the model dies.

File: components/eam/src/physics/crm/pam/pam_driver.cpp [hunk 5]
```diff
@@ -150,12 +220,15 @@ extern "C" void pam_driver() {
...
+  #ifdef MMF_DISABLE_DENSITY_RECALL
+    do_density_save_recall = false;
+  #endif
   //------------------------------------------------------------------------------------------------
   //------------------------------------------------------------------------------------------------
   //------------------------------------------------------------------------------------------------
 
   // set number of CRM steps
-  int nstop = int(gcm_dt/crm_dt);
+  int nstop = int(gcm_dt/crm_dt_fixed);
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_driver.cpp:231]: This has an implied `floor` operation. Is that what you intend if they aren't evenly divisible? Another way to handle non-evenly divisible could be:
```C++
int nstop = std::ceil(gcm_dt / crm_dt_fixed);
crm_dt_fixed = gcm_dt / nstop
```

File: components/eam/src/physics/crm/pam/pam_state.h [hunk 6]
```diff
@@ -144,7 +153,7 @@ inline void pam_state_set_reference_state( pam::PamCoupler &coupler ) {
   real2d hmean_temp ("hmean_temp" ,nz  ,nens);
   real r_nx_ny  = 1._fp/(nx*ny);  // precompute reciprocal to avoid costly divisions
   // initialize horizontal means
-  parallel_for(SimpleBounds<2>(nz,nens), YAKL_LAMBDA (int k, int iens) {
+  parallel_for(SimpleBounds<2>(nz+1,nens), YAKL_LAMBDA (int k, int iens) {
     hmean_pint(k,iens) = 0;
     if (k < nz) { 
       hmean_pmid (k,iens) = 0;
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_state.h:159]: This is being set to zero here but not atomically added in the next kernel.

File: components/eam/src/physics/crm/pam/pam_variance_transport.h [hunk 7]
```diff
@@ -36,6 +38,8 @@ inline void pam_variance_transport_diagnose( pam::PamCoupler &coupler ) {
   //------------------------------------------------------------------------------------------------
   auto temp         = dm_device.get<real,4>("temp"       );
   auto rhov         = dm_device.get<real,4>("water_vapor");
+  auto rhoc         = dm_device.get<real,4>("cloud_water");
+  auto rhoi         = dm_device.get<real,4>("ice"        );
   auto uvel         = dm_device.get<real,4>("uvel"       );
```
Comment (mrnorman) [inline components/eam/src/physics/crm/pam/pam_variance_transport.h:43]: These should be `real const`, I think.

**Review by mrnorman — APPROVED** [review]

**Comment by rljacob** [comment]
@whannah1 anything else to do here?

File: components/eam/src/physics/crm/pam/pam_driver.cpp (hunk 2 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_driver.cpp:64]: Good call - I can't remember why I did it this way.

File: components/eam/src/physics/crm/pam/pam_variance_transport.h (hunk 7 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_variance_transport.h:43]: Yea, good catch.

File: components/eam/src/physics/crm/pam/pam_state.h (hunk 6 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_state.h:159]: yea, good eye, I'll remove this line.

Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_state.h:159]: fixed.

File: components/eam/src/physics/crm/pam/pam_variance_transport.h (hunk 7 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_variance_transport.h:43]: fixed.

File: components/eam/src/physics/crm/pam/pam_driver.cpp (hunk 4 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_driver.cpp:88]: Cool, thanks for the tip!

File: components/eam/src/physics/crm/pam/pam_driver.cpp (hunk 3 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_driver.cpp:76]: Excellent, just made the fix.

File: components/eam/src/physics/crm/pam/pam_driver.cpp (hunk 5 above)
Comment (whannah1) [reply components/eam/src/physics/crm/pam/pam_driver.cpp:231]: I'm actually not sure which is better or if it even matters.... so I'm just gonna leave the implied floor in there.

**Comment by whannah1** [comment]
@rljacob nothing left to do here - this is ready to merge whenever someone can get to it.

**Comment by whannah1** [comment]
@rljacob looks like there's a new problem that's been created by other updates - which I think will all be fixed when the eamxx code in this repo is updated. Not sure how to navigate this, so I might just leave this in a holding pattern for a bit.

**Comment by rljacob** [comment]
Waiting for SCREAM update #6489 and a few more changes.

**Comment by rljacob** [comment]
@whannah1 is this ready?

**Comment by whannah1** [comment]
@rljacob @brhillman This PR is ready to go.

**Comment by brhillman** [comment]
Merged to next.

**Pull request merged**